File: Hashing/csharp/PairSum.cs
using System;
using System.Collections.Generic;

namespace HashingSolutions;

public static class PairSum
{
    public static (int, int) Solve(IReadOnlyList<int> numbers, int targetSum)
    {
        var indices = new Dictionary<int, int>();
        for (var i = 0; i < numbers.Count; i++)
        {
            var value = numbers[i];
            var complement = targetSum - value;
            if (indices.TryGetValue(complement, out var index))
            {
                return (index, i);
            }

            if (!indices.ContainsKey(value))
            {
                indices[value] = i;
            }
        }

        throw new InvalidOperationException("Pair not found");
    }
}

File: Hashing/java/pairSum.java
import java.util.HashMap;
import java.util.Map;

class PairSum {
  public static int[] pairSum(int[] numbers, int targetSum) {
    Map<Integer, Integer> indices = new HashMap<>();
    for (int i = 0; i < numbers.length; i += 1) {
      int value = numbers[i];
      int difference = targetSum - value;
      if (indices.containsKey(difference)) {
        return new int[] { indices.get(difference), i };
      }
      indices.putIfAbsent(value, i);
    }
    throw new IllegalArgumentException("Pair not found");
  }

  public static void run() {
    // intentionally empty
  }
}
